Added widget header refresh indicator

ranbena opened the pull request:
- [x] Feature

## Description
Testing indicator design.
Play with it [in the preview](https://deploy-preview-3970--redash-preview.netlify.com/dashboard/artists?p_AlbumId=4&p_test=MrCirca&p_text=Wor&p_number=4&p_date=2019-06-05&p_range.start=2019-06-07&p_range.end=2019-06-21&p_w191_AlbumId=4&p_w210_new=test).

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/486954/60990251-ac017a80-a350-11e9-9c56-d0ad22727651.gif)

## Related Tickets & Documents
https://github.com/getredash/redash/issues/3878

arikfr commented:
It looks good, but why in the right top corner and not in the bottom left where we already have a spinner? Just replace the existing spinner with this one when the query executes?

ranbena commented:
> It looks good, but why in the right top corner and not in the bottom left where we already have a spinner? Just replace the existing spinner with this one when the query executes?

Our assumption is that it's not clear when a widget is refreshing. It's obvious that the text and spinner are inconspicuous, so that's what I went for with my first experimentations.
But I concluded it's still not clear enough since lower/tall widgets had their bottom part obscured, hiding the indication - so I added the indication up top.
Also, conceptually this type of information feels to me like the "front and center" kind.

I can make the bottom only version to get the feel and help decide.

arikfr commented:
> But I concluded it's still not clear enough since lower/tall widgets had their bottom part obscured, hiding the indication - so I added the indication up top.

Good point.

It just bothers me that we have multiple spinners and counters. Maybe we should move the counter next to the spinner at the top?

ranbena commented:
> It just bothers me that we have multiple spinners and counters.

Yeah, it's now 3 corners with refresh spinners :D

> Maybe we should move the counter next to the spinner at the top?

Depends what this info is for - to gain a rough estimate of how much longer it's gonna take? To indicate a stuck job?

arikfr commented:
> Depends what this info is for - to gain a rough estimate of how much longer it's gonna take? To indicate a stuck job?

Very deep questions for 11PM ... :-)

Another option is to drop the spinner next to the counter, but keep the counter where it currently is.

ranbena commented:
> Another option is to drop the spinner next to the counter, but keep the counter where it currently is.

Why not show the timer next to the spinner?
1. It's not valuable information unless job stuck.
2. It'll take up widget title real-estate.

Proposed UX:
1. Remove the bottom spinner and timer.
2. Show timer in tooltip when hovering the spinner.
3. When job stuck, change spinner color to red and link to troubleshooting doc from tooltip.

arikfr commented:
Ok, let's give a try to:

> Remove the bottom spinner and timer.
> Show timer in tooltip when hovering the spinner.

(we have no concept of "stuck", so we can skip this for now)

ranbena commented:
> Ok, let's give a try to:
> 
> > Remove the bottom spinner and timer.
> > Show timer in tooltip when hovering the spinner.
> 
> (we have no concept of "stuck", so we can skip this for now)

Actually, it's a bit more complex than that. The bottom refresh indicator has two more features that can't be easily ported to the top:
1. Time ago info - important to have at a glance. So can't be put it in a tooltip, nor persist in the widget header.
2. Acts as a refresh button - but top indicator was designed to be an indication only, not an interactive component (moves away and resizes on hover).

I might need to make some more changes.

arikfr commented:
> Time ago info - important to have at a glance. So can't be put it in a tooltip, nor persist in the widget header.

I thought of keeping only this, and hiding it when the query executes.

ranbena commented:
> I thought of keeping only this, and hiding it when the query executes.

It doesn't make sense to me. You press a button, it disappears, and an indication shows up at the opposite corner (which might be off-screen).

ranbena commented:
I'm trying a different approach - a candy stripe animation atop but my current implementation is cpu intensive and doesn't work well.

![ezgif com-video-to-gif (1)](https://user-images.githubusercontent.com/486954/61125039-52fd2800-a4b1-11e9-8fa3-a3e4c92e4d5f.gif)

But lmk if you think this has potential.

ranbena commented:
Oh and also, if the run time is under an hour I display `mm:ss` instead of `HH:mm:ss`.

arikfr commented:
> It doesn't make sense to me. You press a button, it disappears, and an indication shows up at the opposite corner (which might be off-screen).

It shows when the query was executed, so in a way it makes sense that you won't see it when you're currently executing the query.

> I'm trying a different approach - a candy stripe animation atop but my current implementation is cpu intensive and doesn't work well.

I think it has potential. Mainly because it works well with the existing elements. The downside is that it feels a bit out of touch with the rest of the visual language of the application.
Maybe we can use an image animation to reduce the CPU overhead?

ranbena commented:
Reverted candy stripe and updated the following changes:

1. Bottom refresh button now shows "last updated" persistently and refresh runtime counter moved to top on hover. Also, removed `HH` when runtime is under 1 hour.
   
   <img src="https://user-images.githubusercontent.com/486954/61967211-79f14900-afdd-11e9-8e25-f6efc251d4c8.gif" width=200 />

3. It was agreed that the bottom spinner adds too much visual noise and that one spinning icon would suffice. So I halted it.
But this resulted in  no clear indication that the refresh has started when clicking the two bottom widget refresh buttons. 
My solution is to have the clicked icon spin till the refresh cycle is done, or the other button gets clicked. All other scenarios they stay still.
I think it strikes the right balance between usability and visual noise.

   <img src="https://user-images.githubusercontent.com/486954/61991510-af3b7c80-b059-11e9-8bda-290031a2ab5f.gif" width=400 />

4. Since the timer indication could potentially overlap a long query title, added a semi-transparent underlay.

   <img width="200" alt="Screen Shot 2019-07-27 at 10 25 30" src="https://user-images.githubusercontent.com/486954/61991456-f1b08980-b058-11e9-87a2-4add7578f471.png">

ranbena commented:
Also, added a simple dashboard "🔄Refresh" button indication.

<img width=200 src="https://user-images.githubusercontent.com/486954/61992113-9388a400-b062-11e9-8776-efd878263e86.gif" />

arikfr commented:
> My solution is to have the clicked icon spin till the refresh cycle is done, or the other button gets clicked. All other scenarios they stay still.

The inconsistency feels confusing. Maybe they shouldn't animate at all? In most cases they will see the top indicator, so it's clear refresh started.

Also on Ubuntu/Chrome the refresh indicator looks unaligned:
![image](https://user-images.githubusercontent.com/71468/62005167-cd77aa00-b137-11e9-9717-c17f3a740d11.png)

ranbena commented:
> The inconsistency feels confusing. Maybe they shouldn't animate at all? In most cases they will see the top indicator, so it's clear refresh started.

If you played around with it and feel it's inconsistent and visually cluttered - I'm convinced.
But if not - plz do, the experience is the true test of this feature.

> Also on Ubuntu/Chrome the refresh indicator looks unaligned:

👍Thanks.

arikfr commented:
> If you played around with it and feel it's inconsistent and visually cluttered - I'm convinced.

I did play with it, but don't forget I'm biased: I already know what should happen...

ranbena commented:
> I did play with it, but don't forget I'm biased: I already know what should happen...

Good enough for me.
I revoked the button click spin in ab5494a. But keeping it on the dashboard refresh button.

@arikfr lmk if this feature is complete as far as you're concerned and I'll release it from draft.

arikfr commented on client/app/components/dashboards/widget.html:
```diff
@@ -29,6 +29,12 @@
             <li ng-if="$ctrl.dashboard.canEdit()"><a ng-click="$ctrl.deleteWidget()">Remove from Dashboard</a></li>
           </ul>
         </div>
+        <div id="topRefreshIndicator" ng-if="$ctrl.widget.loading">
```
Isn't `id` supposed to be singular on a page?

ranbena replied:
Only when retrieving elements by id but in terms of css hooks, it has no significance.

I'll change it though 👍🏻

kravets-levko commented on client/app/components/Timer.jsx:
```diff
@@ -14,7 +14,10 @@ export function Timer({ from }) {
 
-  return moment.utc(moment.now() - startTime).format('HH:mm:ss');
+  const diff = moment.now() - startTime;
+  const format = diff > 1000 * 60 * 60 ? 'HH:mm:ss' : 'mm:ss'; // no HH under an hour
```
:+1:

kravets-levko commented on client/app/assets/less/redash/redash-newstyle.less:
```diff
@@ -480,6 +490,86 @@ body {
   background-color: fade(@redash-gray, 12%) !important;
 }
 
+#topRefreshIndicator {
```
Kebab-case please :pray: Also, don't use `id` for elements that may occur multiple times on page - they're not intended for this. Also, ids are very hard to override in CSS because they have highest priority and support the only value in attribute.

ranbena replied:
40ec0c0

arikfr commented on client/app/assets/less/redash/redash-newstyle.less:
```diff
@@ -480,6 +490,86 @@ body {
   background-color: fade(@redash-gray, 12%) !important;
 }
 
+.refresh-indicator {
```
Something that I missed when doing the review: we need to stop adding styles to `redash-newstyle.less`. The refresh indicator should've been a component and its style should've been next to it.

Not worth revisiting at the moment, just wanted to point out so we keep this in mind for future stuff.

ranbena replied:
Had to be done! https://github.com/getredash/redash/pull/4017

arikfr commented:
@ranbena ,

> I did play with it, but don't forget I'm biased: I already know what should happen...

Now that I had real life experience with it, I can say that I was wrong. When you click on a widget to refresh, it's in the more subtle mode of the refresh indicator. Therefore it's easy to miss, and because the button doesn't animate it feels like nothing happens.

I guess that we should bring back the animation on the button you click you had there...

ranbena commented:
I had an alternative I tried - making the buttons have a hover+press+pressed state.
Still, it's enough without the icon spinning.

<img src="https://user-images.githubusercontent.com/486954/62245738-e8257980-b3ea-11e9-8b2b-77c3f566100a.gif" width=500 />

ranbena commented:
Reverted in https://github.com/getredash/redash/pull/4027
